pkgs/native_assets_cli/lib/src/c_compiler_config.dart

```diff
@@ -0,0 +1,129 @@
+// Copyright (c) 2023, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'package:collection/collection.dart';
+
+import 'json_utils.dart';
+import 'utils/map.dart';
+
+/// The configuration for a C toolchain.
+final class CCompilerConfig {
+  /// Path to a C compiler.
+  late final Uri? compiler;
+
+  /// Path to a native linker.
+  late final Uri? linker;
+
+  /// Path to a native archiver.
+  late final Uri? archiver;
+
+  /// Path to script that sets environment variables for [compiler], [linker],
+  /// and [archiver].
+  late final Uri? envScript;
+
+  /// Arguments for [envScript].
+  late final List<String>? envScriptArgs;
+
+  /// Constructs a new [CCompilerConfig] based on the given toolchain tools.
+  CCompilerConfig({
+    this.archiver,
+    this.compiler,
+    this.linker,
+    this.envScript,
+    this.envScriptArgs,
+  });
+
+  /// Constructs a [CCompilerConfig] from the given [json].
+  ///
+  /// The json is expected to be valid encoding obtained via
+  /// [CCompilerConfig.toJson].
+  factory CCompilerConfig.fromJson(Map<String, Object?> json) {
+    final compiler = _parseCompiler(json);
+    return CCompilerConfig(
+      archiver: _parseArchiver(json),
+      compiler: compiler,
+      envScript: _parseEnvScript(json, compiler),
+      envScriptArgs: _parseEnvScriptArgs(json),
+      linker: _parseLinker(json),
+    );
+  }
+
+  // TODO(https://github.com/dart-lang/native/issues/1599): The main reason
+  // these keys are exposed is due to usage on Dart CI testing infrastructure.
+  // If our infrastructure supplies the native toolchains via passing those down
+  // to the `dart build/run` via overrides we should no longer need to expose
```

Comment:
We also run the `native_toolchain_c` tests on the Dart CI. So it's not only the `dart build/run` for which this needs to work.
(The reason we run those tests on the Dart CI is that we use very native compilers (and different versions) there than on the GitHub CI, and having only the dartdev tests fail makes it hard to see if there's a bug in dartdev or a bug in native_toolchain_c which only surfaced with the native toolchains on the Dart CI.)

Reply:
Thanks. Updated comment.